new post: "Unlocking Universal Libraries in Dune"

anmonteiro opened the pull request:
adding a new blog post describing our path towards the solution we're experimenting with re: multi-context libraries

jchavarri reviewed (commented):
Thanks for writing this! It's a great summary of the process we've followed I think.

It might be that the contents of the post end up being deprecated with newer approaches (an example can be found in the inline comments), but I like the naming with "preview" in the file name, maybe it'd be a good idea to mention "preview" somewhere in the post title to give readers the idea that all this territory is under heavy exploration at the moment.

davesnx commented on blog/posts/dune-universal-libraries-preview.md:
```diff
@@ -0,0 +1,241 @@
+## The Bird's-eye View
+
+Let's walk backwards from our end-goal: having a shared OCaml / Melange
+codebase that can render React.js components on the server, such that the
```
I don't consider universal to only render components, but also work with encoding/decoding JSON, share types and validations.

anmonteiro replied:
yeah, me too. In this example, however, I'm walking backwards from a concrete use case, not defining universal libraries.

davesnx commented on blog/posts/dune-universal-libraries-preview.md:
````diff
@@ -0,0 +1,241 @@
+corresponding source files can live in a single directory, no copying required.
+If we need to vary `C`'s implementation, we can express that in Dune rules:
+
+```clj
````
love the clj annotation

anmonteiro commented on blog/posts/dune-universal-libraries-preview.md:
```diff
@@ -0,0 +1,241 @@
+(copy_files# ../native
+ (files {a,b}.ml{,i}))
+
+;; module `C` has a specific Melange implementation
+(rule
+ (with-stdout-to c.ml
+  (echo "let backend = \"melange\"")))
```
> I think the key part missing in this example dune config block is the native/shared folder where we copy_files from. Then one can use include_subdirs unqualified, and put platform specific modules in either native or melange.

Isn't this satisfied by the `copy_files#` stanzas below?

anmonteiro replied:
because creating a rule makes it explicit that it's a melange-specific and native-specific module. The alternative would be to describe an external dune file, but this way we can just show dune files and nothing else.

jchavarri replied:
Makes sense.

anmonteiro commented on blog/posts/dune-universal-libraries-preview.md:
````diff
@@ -0,0 +1,241 @@
+If we need to vary `C`'s implementation, we can express that in Dune rules:
+
+```clj
+(rule
+ (target c.ml)
+ (deps c.native.ml)
+ (action
+  (with-stdout-to
+   %{target}
+   (echo "let backend = \"OCaml\"")))
+ (enabled_if
+  (= %{context_name} default)))
+
+(rule
+ (target c.ml)
+ (deps c.melange.ml)
+ (action
+  (with-stdout-to
+   %{target}
+   (echo "let backend = \"Melange\"")))
+ (enabled_if
+  (= %{context_name} melange)))
````
I'd rather keep it simple while we're under "preview" (I agree with your other comment that we should say this is a preview).

The goal of this post is to tell the story rather than prescribe a dune build setup; I'd love to follow up with a complete tutorial / part 2 once we get it right.

anmonteiro replied:
how about I add a paragraph mentioning just that?

that it's an illustrative example that we don't know whether it's going to work

jchavarri replied:
Would be helpful 👍

anmonteiro commented:
I addressed your comments:

- renamed the post to "A Preview of Universal Libraries in Dune"
- addressed most of the suggestions
- on the code snippets, I have a slight, loosely held disagreement on what we should show. I tried to go for examples that are easiest to understand rather than focusing on specific platform details or introducing extra concepts. I hope this is sufficient to give an idea of what we're working on rather than prescribe specific solutions, which IMO have their place in a tutorial or proper documentation. what do you tink?
